mono/metadata/handle-0.h

```diff
@@ -0,0 +1,141 @@
+#define MONO_TYPE_SAFE_HANDLES 1
+#else
+#define MONO_TYPE_SAFE_HANDLES 0 // PowerPC, S390X, SPARC, MIPS, Linux/x86, BSD/x86, etc.
+#endif
+
+/*
+Handle macros/functions
+*/
+
+#define TYPED_HANDLE_PAYLOAD_NAME(TYPE) TYPE ## HandlePayload
+#define TYPED_HANDLE_NAME(TYPE) TYPE ## Handle
+#define TYPED_OUT_HANDLE_NAME(TYPE) TYPE ## HandleOut
+
+// internal helpers:
+#define MONO_HANDLE_CAST_FOR(type) mono_handle_cast_##type
+#define MONO_HANDLE_TYPECHECK_FOR(type) mono_handle_typecheck_##type
+
+/*
+ * TYPED_HANDLE_DECL(SomeType):
+ *   Expands to a decl for handles to SomeType and to an internal payload struct.
+ *
+ * For example, TYPED_HANDLE_DECL(MonoObject) (see below) expands to:
+ *
+ * #if MONO_TYPE_SAFE_HANDLES
+ *
+ * typedef struct {
+ *   MonoObject **__raw;
+ * } MonoObjectHandlePayload,
+ *   MonoObjectHandle,
+ *   MonoObjectHandleOut;
+ *
+ * Internal helper functions are also generated.
+ *
+ * #else
+ *
+ * typedef struct {
+ *   MonoObject *__raw;
+ * } MonoObjectHandlePayload;
+ *
+ * typedef MonoObjectHandlePayload* MonoObjectHandle;
+ * typedef MonoObjectHandlePayload* MonoObjectHandleOut;
+ *
+ * #endif
+ */
+
+#ifdef __cplusplus
+#define MONO_IF_CPLUSPLUS(x) x
+#else
+#define MONO_IF_CPLUSPLUS(x) /* nothing */
+#endif
+
+#if MONO_TYPE_SAFE_HANDLES
+#define TYPED_HANDLE_DECL(TYPE)							\
+	typedef struct {							\
+		MONO_IF_CPLUSPLUS (						\
+			MONO_ALWAYS_INLINE					\
+			TYPE * GetRaw () { return __raw ? *__raw : NULL; }	\
+		)								\
+		TYPE **__raw;							\
+	} TYPED_HANDLE_PAYLOAD_NAME (TYPE),					\
+	  TYPED_HANDLE_NAME (TYPE),						\
+	  TYPED_OUT_HANDLE_NAME (TYPE);						\
+/* Do not call these functions directly. Use MONO_HANDLE_NEW and MONO_HANDLE_CAST. */ \
+/* Another way to do this involved casting mono_handle_new function to a different type. */ \
+static inline MONO_ALWAYS_INLINE TYPED_HANDLE_NAME (TYPE) 	\
+MONO_HANDLE_CAST_FOR (TYPE) (gpointer a)			\
+{								\
+	TYPED_HANDLE_NAME (TYPE) b = { (TYPE**)a };		\
+	return b;						\
+}								\
+static inline MONO_ALWAYS_INLINE MonoObject* 			\
+MONO_HANDLE_TYPECHECK_FOR (TYPE) (TYPE *a)			\
+{								\
+	return (MonoObject*)a;					\
+}
+
+#else
+#define TYPED_HANDLE_DECL(TYPE)						\
+	typedef struct { TYPE *__raw; } TYPED_HANDLE_PAYLOAD_NAME (TYPE) ; \
+	typedef TYPED_HANDLE_PAYLOAD_NAME (TYPE) * TYPED_HANDLE_NAME (TYPE); \
+	typedef TYPED_HANDLE_PAYLOAD_NAME (TYPE) * TYPED_OUT_HANDLE_NAME (TYPE)
+#endif
+
+/*
+ * TYPED_VALUE_HANDLE_DECL(SomeType):
+ *   Expands to a decl for handles to SomeType (which is a managed valuetype (likely a struct) of some sort) and to an internal payload struct.
+ * It is currently identical to TYPED_HANDLE_DECL (valuetypes vs. referencetypes).
+ */
+#define TYPED_VALUE_HANDLE_DECL(TYPE) TYPED_HANDLE_DECL(TYPE)
+
+/* Safely access System.Reflection.Assembly from native code */
+TYPED_HANDLE_DECL (MonoReflectionAssembly);
```

Comment:
I agree it is a wierd place, but consider, like, object-forward.h, and that the types are in object-internal.h or such.
i.e. maybe it is ok for them to be broadly visible?
And as  you indicated, I didn't want to move the function declarations, but I can try that.
handle-forward.h?
I can look again.

Reply:
`.h` mostly but not entirely imples `decl`, so not a great name.
This file is mostly devoid of declarations, has mainly macros/definitions, that, granted, generate declarations or, really, definitions.

`0` is an indication of ordering or graph depth or split into pieces.

Generally there is a graph of types and subsystems/functions, that we mildly attempt to mimic as an include graph, with no enforcement. The file basis of build systems/compilers is really unfortunate.

I'd kinda like this to be in object-internals.h, but that might not work or be more difficult, and loses the informal "handle" / "handle.h" relationship that handle.h and handle-0.h has.

handle-internal.h doesn't work, because handle.h is already internal.

handletypes.h or handledef.h kinda fit, however, I ultimately split off only a very specific fairly minimal chunk, not all types or defines, so rejected those.

I considered typedhandle.h, but that is really what handle.h is short for.

We might consider handle-inlines.h, for the rest of the file. If that works.
But that might require more reordering/restructuring.

handle-forward.h really is tempting -- it'd be like object-forward.h, and have all/some TYPED_HANDLE_DECL, and the definition of TYPED_HANDLE_DECL to support it? Or maybe that is two pieces where only one should be.

Reply:
Perhaps I can reorder the file more or something. I did try. Nothing is very satisfactory here.
It is too bad we don't just have a database of program pieces -- types, functions, etc., and have to group them into files with connective includes. :(

Reply:
I suspect this is a losing battle and we'll want to forward declare more and more structs and handles, but for today, reflection-internals.h won.